llvm/lib/IR/Attributes.cpp

```diff
@@ -903,6 +916,98 @@ AttributeSet AttributeSet::removeAttributes(LLVMContext &C,
   return get(C, B);
 }
 
+std::optional<AttributeSet>
+AttributeSet::intersectWith(LLVMContext &C, AttributeSet Other) const {
+  if (*this == Other)
+    return *this;
+  AttrBuilder Intersected(C);
+  // Get common set of all attributes then handle each attr according to its
+  // intersect rule.
+  AttributeSet Merged = addAttributes(C, Other);
```

Comment:
had been nice to have some better way to get all AttrKinds looks like a lot of extra work to make a new AttributeSet only to get a way to iterate over the AttrKinds.

Reply:
It's possible to do this by iterating over both AttributeSets at once. They are sorted, so you would increment either the left or the right iterator depending on which has the larger kind.